pydantic/_internal/_dataclasses.py

```diff
@@ -98,9 +99,15 @@ def __init__(__dataclass_self__: PydanticDataclass, *args: Any, **kwargs: Any) -
     get_core_schema = getattr(cls, '__get_pydantic_core_schema__', None)
     try:
         if get_core_schema:
-            schema = get_core_schema(cls, partial(gen_schema.generate_schema, from_dunder_get_core_schema=False))
+            schema = get_core_schema(
+                cls,
+                CallbackGetCoreSchemaHandler(
+                    partial(gen_schema.generate_schema_for_type, from_dunder_get_core_schema=False),
+                    gen_schema.generate_schema,
+                ),
+            )
```

Comment:
We fixed this bug for BaseModel recently, just neglected to do so for dataclasses